### PR TITLE
XRT-1981 Schedule serialization support backport

### PR DESCRIPTION
**Pull request by SteveOss**

### CODE REVIEW

File: src/c/scheduler.c [hunk 1]
```diff
@@ -163,31 +170,38 @@ static void * iot_scheduler_thread (void * arg)
...
-      if (current->run_cb) current->run_cb (current->arg);
-      /* Post the work to the thread pool or run as thread */
-      iot_schedule_add_ref (current);
-      if (current->threadpool)
+      if (current->concurrent || (atomic_load (&current->refs) == 1u)) // Check for concurrency
       {
-        iot_log_trace (scheduler->logger, "Running schedule #%" PRIu64 " from threadpool", current->id);
-        if (! iot_threadpool_try_work (current->threadpool, schedule_fn, current, current->priority))
+        /* Notify that the schedule is about to run */
+        if (current->run_cb) current->run_cb (current->arg);
```
Comment (dpazj) [inline src/c/scheduler.c:176]: Do we need to unlock the scheduler component around the callback?

File: src/c/scheduler.c [hunk 2]
```diff
@@ -163,31 +170,38 @@ static void * iot_scheduler_thread (void * arg)
...
-          /* Notify that the run is aborted */
-          if (current->abort_cb) current->abort_cb (current->arg);
-          if (atomic_fetch_add (&current->dropped, 1u) == 0)
+          iot_log_trace (scheduler->logger, "Running schedule #%" PRIu64 " from threadpool", current->id);
+          if (!iot_threadpool_try_work (current->threadpool, schedule_fn, current, current->priority))
           {
-            iot_log_warn (scheduler->logger, "Scheduled event dropped for schedule #%" PRIu64, current->id);
+            iot_schedule_dec_ref (current);
+            /* Notify that the run is aborted */
+            if (current->abort_cb) current->abort_cb (current->arg);
```
Comment (dpazj) [inline src/c/scheduler.c:186]: Do we need to unlock the scheduler component around the abort callback?

**Review by dpazj — COMMENTED** [review]
LGTM. Just a few comments about unlocking the scheduler component which is different to changes in the other versions.

File: src/c/scheduler.c (hunk 2 above)
Comment (SteveOss) [reply src/c/scheduler.c:186]: @edward-scott What do you think ? We must have added the unlock before callback to fix something at some point ?

**Review by edward-scott — APPROVED** [review]
LGTM.

**Comment by SteveOss** [comment]
@edward-scott Do you think we also need to backport the unlock/lock calls around the callbacks ?

File: src/c/scheduler.c (hunk 2 above)
Comment (edward-scott) [reply src/c/scheduler.c:186]: Actually, yes as the abort callback in devsdk can call schedule delete. So both run and abort should unlock.

**Pull request merged**